src/warc2zim/content_rewriting.py

```diff
@@ -0,0 +1,168 @@
+    return attr
+
+
+def format_attr(name: str, value: Optional[str]) -> str:
+    if value is None:
+        return name
+    html_escaped_value = escape(value, quote=True)
+    return f'{name}="{html_escaped_value}"'
+
+
+def transform_attrs(
+    attrs: AttrsList, url_rewriter: Callable[[str], str], css_rewriter: "CssRewriter"
+) -> str:
+    processed_attrs = (process_attr(attr, url_rewriter, css_rewriter) for attr in attrs)
+    return " ".join(format_attr(*attr) for attr in processed_attrs)
+
+
+class RewriterParser(HTMLParser):
+    def __init__(
+        self, article_url: str, pre_head_insert: str, post_head_insert: str, sink
+    ):
+        super().__init__()
+        self.url_rewriter = ArticleUrlRewriter(article_url)
+        self.css_rewriter = CSSRewriter(article_url)
+        self.sink = sink
+        self.title = None
+        # This works only for tag without children.
+        # But as we use it to get the title, we are ok
+        self._active_tag = None
+        self.pre_head_insert = pre_head_insert
+        self.post_head_insert = post_head_insert
+
+    def send(self, value: str):
+        self.sink.write(value)
+
+    def handle_starttag(self, tag: str, attrs: AttrsList, auto_close: bool = False):
+        self._active_tag = tag
+
+        self.send(f"<{tag}")
+        if attrs:
+            self.send(" ")
+        self.send(transform_attrs(attrs, self.url_rewriter, self.css_rewriter))
+
+        if auto_close:
+            self.send(" />")
+        else:
+            self.send(">")
+        if tag == "head" and self.pre_head_insert:
+            self.send(self.pre_head_insert)
+
+    def handle_endtag(self, tag: str):
+        self._active_tag = None
+        if tag == "head" and self.post_head_insert:
+            self.send(self.post_head_insert)
+        self.send(f"</{tag}>")
+
+    def handle_startendtag(self, tag: str, attrs: AttrsList):
+        self.handle_starttag(tag, attrs, auto_close=True)
+        self._active_tag = None
+
+    def handle_data(self, data: str):
+        if self._active_tag == "title" and self.title is None:
+            self.title = data.strip()
+        elif self._active_tag == "style":
+            data = self.css_rewriter.rewrite(data)
+        self.send(data)
+
+    def handle_comment(self, data: str):
+        self.send(f"<!--{data}-->")
+
+    def handle_decl(self, decl: str):
+        self.send(f"<!{decl}>")
+
+    def handle_pi(self, data: str):
+        self.send(f"<?{data}>")
+
+    def unknown_decl(self, data: str):
+        self.handle_decl(data)
+
+
+def rewrite_html(
+    content: str | bytes, article_url: str, pre_head_insert: str, post_head_insert: str
+) -> tuple[str, str]:
+    content = to_string(content)
+    output = io.StringIO()
+
+    parser = RewriterParser(article_url, pre_head_insert, post_head_insert, output)
+    parser.feed(content)
+    parser.close()
+
+    return (parser.title or "", output.getvalue())
```

Comment:
Returning a namedtuple might be useful ; I see you're not calling it much in code but you do in tests